### PR TITLE
Change motor control input from Amps to Nm

### PR DESCRIPTION
**Pull request by PAJohnson**
The purpose of this PR is to change from units of Amps to units of Nm for the input value to the motor control. This affects multiple parts of the ODrive code structure, including the communication interfaces.

### CODE REVIEW

File: Firmware/MotorControl/controller.cpp [hunk 1]
```diff
@@ -153,29 +153,29 @@ bool Controller::update(float* current_setpoint_output) {
...
         case INPUT_MODE_VEL_RAMP: {
             float max_step_size = std::abs(current_meas_period * config_.vel_ramp_rate);
             float full_step = input_vel_ - vel_setpoint_;
             float step = std::clamp(full_step, -max_step_size, max_step_size);
 
             vel_setpoint_ += step;
-            current_setpoint_ = (step / current_meas_period) * config_.inertia;
+            torque_setpoint_ = (step / current_meas_period) * config_.inertia;
         } break;
         case INPUT_MODE_CURRENT_RAMP: {
```
Comment (madcowswe) [inline Firmware/MotorControl/controller.cpp]: We should also rename this to Torque ramp.

File: Firmware/MotorControl/motor.cpp [hunk 2]
```diff
@@ -177,19 +177,19 @@ bool Motor::do_checks() {
     return true;
 }
 
-float Motor::effective_current_lim() {
+float Motor::effective_torque_lim() {
```
Comment (madcowswe) [inline Firmware/MotorControl/motor.cpp]: Hm actually I think all the limits involved here make more sense to be in terms of current. I would keep this function as effective_current_lim, and apply the conversion at every applicable call site.

File: Firmware/MotorControl/motor.hpp [hunk 3]
```diff
@@ -29,6 +29,7 @@ class Motor : public ODriveIntf::MotorIntf {
         float Id_measured; // [A]
         float I_measured_report_filter_k;
         float max_allowed_current; // [A]
+        float max_allowed_torque; // [Nm]
```
Comment (madcowswe) [inline Firmware/MotorControl/motor.hpp]: "max allowed current" is a hardware limit, as noted earlier, probably best to keep the current limits as current and not torque.

File: Firmware/communication/ascii_protocol.cpp [hunk 4]
```diff
@@ -108,15 +108,15 @@ void ASCII_protocol_process_line(const uint8_t* buffer, size_t len, StreamSink&
...
             axis->controller_.input_pos_updated();
             axis->watchdog_feed();
         }
 
     } else if (cmd[0] == 'q') { // position control with limits
         unsigned motor_number;
-        float pos_setpoint, vel_limit, current_lim;
-        int numscan = sscanf(cmd, "q %u %f %f %f", &motor_number, &pos_setpoint, &vel_limit, &current_lim);
+        float pos_setpoint, vel_limit, torque_lim;
+        int numscan = sscanf(cmd, "q %u %f %f %f", &motor_number, &pos_setpoint, &vel_limit, &torque_lim);
```
Comment (madcowswe) [inline Firmware/communication/ascii_protocol.cpp:119]: Here is a good example where we do in fact want a torque limit (not current limit), as it's a proxy for an acceleration limit. I guess we should have both a torque and current limit be observed seperately.

File: docs/getting-started.md [hunk 5]
```diff
@@ -346,9 +346,9 @@ Set the velocity ramp rate (acceleration): `axis.controller.config.vel_ramp_rate
 Activate the ramped velocity mode: `axis.controller.config.input_mode = INPUT_MODE_VEL_RAMP`.<br>
 You can now control the velocity with `axis.controller.input_vel = 5000` [count/s].
 
-### Current control
+### Torque control
 Set `axis.controller.config.control_mode = CONTROL_MODE_CURRENT_CONTROL`.<br>
-You can now control the current with `axis.controller.input_current = 3` [A].
+You can now control the torque with `axis.controller.input_torque = 3` [Nm].
```
Comment (madcowswe) [inline docs/getting-started.md]: 3Nm is very large (comared to 3 A on common motors). Perhaps use 0.1Nm in the example.

File: tools/.vscode/launch.json [hunk 6]
```diff
@@ -9,7 +9,7 @@
             "type": "python",
             "request": "launch",
             "stopOnEntry": true,
-            "pythonPath": "${config:python.pythonPath}",
+            "pythonPath": "${config:python.interpreterPath}",
```
Comment (madcowswe) [inline tools/.vscode/launch.json]: Is this change intended? What is the reason it is changed (I don't know what either does)

File: Firmware/MotorControl/motor.cpp [hunk 7]
```diff
@@ -440,13 +440,14 @@ bool Motor::FOC_current(float Id_des, float Iq_des, float I_phase, float pwm_pha
...
+    float current_setpoint = torque_setpoint / config_.torque_constant;
     current_setpoint *= config_.direction;
     phase *= config_.direction;
     phase_vel *= config_.direction;
 
     // TODO: 2-norm vs independent clamping (current could be sqrt(2) bigger)
-    float ilim = effective_current_lim();
+    float ilim = effective_torque_lim() / config_.torque_constant;
     float id = std::clamp(current_control_.Id_setpoint, -ilim, ilim);
     float iq = std::clamp(current_setpoint, -ilim, ilim);
```
Comment (madcowswe) [inline Firmware/MotorControl/motor.cpp]: Note that `iq = torque_setpoint / torque_constant` doesn't work for ACIM motors. In that case you need to do:
`iq = torque_setpoint / (torque_constant * acim_rotor_flux)`. Beware of dividing by zero acim_rotor_flux, you can use `max(acim_rotor_flux, acim_gain_min_flux)`

File: Firmware/MotorControl/controller.cpp (hunk 1 above)
Comment (PAJohnson) [reply Firmware/MotorControl/controller.cpp]: I think you meant this in reference to INPUT_MODE_CURRENT_RAMP. I changed that to INPUT_MODE_TORQUE_RAMP.

File: tools/.vscode/launch.json (hunk 6 above)
Comment (PAJohnson) [reply tools/.vscode/launch.json]: This change was not intended and I also do not know what it does.

File: Firmware/MotorControl/motor.hpp [hunk 8]
```diff
@@ -159,6 +162,7 @@ class Motor : public ODriveIntf::MotorIntf {
     } gate_driver_exported_;
     DRV_SPI_8301_Vars_t gate_driver_regs_; //Local view of DRV registers (initialized by DRV8301_setup)
     float thermal_current_lim_ = 10.0f;  //[A]
+    float thermal_torque_lim_ = 10.0f; //[Nm]
```
Comment (madcowswe) [inline Firmware/MotorControl/motor.hpp]: This variable is not used.

File: Firmware/MotorControl/motor.cpp [hunk 9]
```diff
@@ -440,11 +440,19 @@ bool Motor::FOC_current(float Id_des, float Iq_des, float I_phase, float pwm_pha
 }
 
 
-bool Motor::update(float current_setpoint, float phase, float phase_vel) {
-    current_setpoint *= config_.direction;
+bool Motor::update(float torque_setpoint, float phase, float phase_vel) {
+    float current_setpoint;
```
Comment (madcowswe) [inline Firmware/MotorControl/motor.cpp]: Maybe set to 0 by default to avoid magic smoke in case of a code error in the future.

File: Firmware/MotorControl/motor.hpp [hunk 10]
```diff
@@ -45,12 +45,15 @@ class Motor : public ODriveIntf::MotorIntf {
         float resistance_calib_max_voltage = 2.0f; // [V] - You may need to increase this if this voltage isn't sufficient to drive calibration_current through the motor.
         float phase_inductance = 0.0f;        // to be set by measure_phase_inductance
         float phase_resistance = 0.0f;        // to be set by measure_phase_resistance
+        float torque_constant = 1.0f;         // to be set by user
```
Comment (madcowswe) [inline Firmware/MotorControl/motor.hpp]: Units: Nm/A for PM motors, Nm/A^2 for induction motors.

File: Firmware/MotorControl/controller.cpp [hunk 11]
```diff
@@ -316,36 +316,35 @@ bool Controller::update(float* current_setpoint_output) {
...
     } else {
         if (limited) {
             // TODO make decayfactor configurable
-            vel_integrator_current_ *= 0.99f;
+            vel_integrator_torque_ *= 0.99f;
         } else {
-            vel_integrator_current_ += ((vel_integrator_gain * gain_scheduling_multiplier) * current_meas_period) * v_err;
+            vel_integrator_torque_ += ((vel_integrator_gain * gain_scheduling_multiplier) * current_meas_period) * v_err;
         }
     }
```
Comment (madcowswe) [inline Firmware/MotorControl/controller.cpp]: Since this no longer considers all kinds of current/torque limits (just the straight torque one), we are not implementing the anti-windup correctly. We need to consider all saturation effects on the "actuator" i.e. torque production.
I will follow up on Discord with some theory.

File: Firmware/MotorControl/motor.cpp [hunk 12]
```diff
@@ -192,6 +192,21 @@ float Motor::effective_current_lim() {
     return current_lim;
 }
 
+float Motor::max_available_torque() {
+    //return the maximum available torque for the motor.
+    //Note - for ACIM motors, available torque is allowed to be 0.
+    if (config_.motor_type == Motor::MOTOR_TYPE_ACIM) {
+        float max_torque = effective_current_lim() * config_.torque_constant * current_control_.acim_rotor_flux;
```
Comment (madcowswe) [inline Firmware/MotorControl/motor.cpp:199]: While not a desired operating point, it can happen that acim_rotor_flux is negative. In that case we probably want to clamp it to 0 to avoid issues with reporting negative max_torque.

File: Firmware/MotorControl/motor.cpp [hunk 13]
```diff
@@ -192,6 +192,21 @@ float Motor::effective_current_lim() {
     return current_lim;
 }
 
+float Motor::max_available_torque() {
+    //return the maximum available torque for the motor.
+    //Note - for ACIM motors, available torque is allowed to be 0.
```
Comment (madcowswe) [inline Firmware/MotorControl/motor.cpp]: By convention we usually put general function documentation right above the function definition rather than right below.
